lalrpop/src/lr1/lane_table/construct/group.rs

```diff
@@ -0,0 +1,165 @@
+        let groups = vec![None; number_of_states];
+        Groups {
+            groups:             groups,
+            context_sets:       Vec::new(),
+            unification_keys:   Vec::new(),
+            unification_table:  UnificationTable::new(),
+        }
+    }
+
+    pub fn allocate(&mut self, state: StateIndex, context_set_opt : Option<&ContextSet>) -> Group {
+        let group = Group { index : self.context_sets.len() };
+        debug!("Groups::allocate: allocated group={:?} for state={:?} context_set_opt={:?}", group, state, context_set_opt);
+        if let Some(context_set) = context_set_opt {
+            self.context_sets.push(context_set.clone())
+        } else {
+            self.context_sets.push(ContextSet::new(0))
+        }
+        self.groups[state.0] = Some(group);
+        self.unification_keys.push(self.unification_table.new_key(group));
+        group
+    }
+
+    pub fn group(&self, state : StateIndex) -> Option<Group> {
+        self.groups[state.0]
+    }
+
+    pub fn add_state(&mut self, state : StateIndex) {
+        assert!(state.0 == self.groups.len());
+        self.groups.push(None)
+    }
+
+    pub fn merge_state(&mut self, group: Group, state : StateIndex, context_set_opt : Option<&ContextSet>)
+        -> bool
+    {
+        self.groups[state.0] = Some(group);
+        if let Some(context_set) = context_set_opt {
+            match self.context_set_mutref(group).inplace_union(context_set) {
+                Err(_) => {
+                    debug!("Group::merge_state: failed to merge group={:?} with state={:?}", group, state);
+                    false
+                } 
+                Ok(_) => {
+                    debug!("Group::merge_state: successfull merge group={:?} with state={:?} context_set={:?}", group, state, self.context_set_ref(group));
+                    true
+                } 
+            }
+        } else {
+            true
+        }
+    }
+
+    fn unify_key(&self, group: Group) -> GroupUnifyKey {
+        self.unification_keys[group.index]
+    }
+
+    fn context_set(&mut self, group: Group) -> ContextSet {
+        let key = self.unify_key(group);
+        let empty = ContextSet::new(0);
+        replace(&mut self.context_sets[self.unification_table.probe_value(key).index], empty)
+    }
+
+    fn context_set_mutref(&mut self, group: Group) -> &mut ContextSet {
+        let key = self.unify_key(group);
+        &mut self.context_sets[self.unification_table.probe_value(key).index]
+    }
+
+    pub fn context_set_ref(&mut self, group: Group) -> &ContextSet {
+        let key = self.unify_key(group);
+        &self.context_sets[self.unification_table.probe_value(key).index]
+    }
+
+    pub fn context_set_ref_for_state(&mut self, state: StateIndex) -> &ContextSet {
+        let group = self.groups[state.0].unwrap().clone();
+        self.context_set_ref(group)
+    }
+
+    pub fn unify_groups(&mut self, group1: Group, group2: Group) {
```

Comment:
There is a place for micro optimization. I am always finding canonical elements for given groups. But we can skip one `UnificationTable::find` call since we know that actually the newest group will be a root group. You mention it in the issue #205 description that we can use this knowledge.

 But I believe skipping `find` on a group which is actually a root group will not give much here.